File: backend/src/cultivars/domain/commands.py
# Standard Library
import uuid
from typing import Annotated

# External Libraries
from pydantic import AfterValidator, BeforeValidator, Field

# VerdanTech Source
from src import settings
from src.common.adapters.utils.spec_manager import SpecManager, Specs
from src.common.domain import Command

from .attributes import CultivarAttributeUpdateCommand
from .enums import CultivarCollectionVisibilityEnum
from .specs import specs

# Load all banned cultivar names and keys
banned_fields = []
with open(settings.static_path("banned_fields.txt"), "r") as file:
    for line in file:
        field = line.strip()
        banned_fields.append(field.lower())

# ======================================
# Fields
# ======================================

CultivarName = Annotated[
    str,
    BeforeValidator(lambda v: v.strip()),
    AfterValidator(SpecManager.get_validation_method(specs, "cultivar_name")),
    # Note: Field used only for annotation, to allow custom error messages.
    Field(
        description=specs.descriptions["cultivar_name"]["field"],
        json_schema_extra={
            "min_length": specs.values["cultivar_name"][Specs.MIN_LENGTH],
            "max_length": specs.values["cultivar_name"][Specs.MAX_LENGTH],
            "pattern": specs.values["cultivar_name"][Specs.PATTERN],
        },
    ),
]
CultivarNames = Annotated[
    list[CultivarName],
    AfterValidator(SpecManager.get_validation_method(specs, "cultivar_names")),
    # Note: Field used only for annotation, to allow custom error messages.
    Field(
        description=specs.descriptions["cultivar_names"]["field"],
        json_schema_extra={
            "max_length": specs.values["cultivar_names"][Specs.MAX_LENGTH],
        },
    ),
]
CultivarKey = Annotated[
    str,
    BeforeValidator(lambda v: v.strip()),
    AfterValidator(SpecManager.get_validation_method(specs, "cultivar_key")),
    # Note: Field used only for annotation, to allow custom error messages.
    Field(
        description=specs.descriptions["cultivar_key"]["field"],
        json_schema_extra={
            "min_length": specs.values["cultivar_key"][Specs.MIN_LENGTH],
            "max_length": specs.values["cultivar_key"][Specs.MAX_LENGTH],
            "pattern": specs.values["cultivar_key"][Specs.PATTERN],
        },
    ),
]
CultivarScientificName = Annotated[
    str,
    BeforeValidator(lambda v: v.strip()),
    AfterValidator(
        SpecManager.get_validation_method(specs, "cultivar_scientific_name")
    ),
    # Note: Field used only for annotation, to allow custom error messages.
    Field(
        description=specs.descriptions["cultivar_scientific_name"]["field"],
        json_schema_extra={
            "max_length": specs.values["cultivar_scientific_name"][Specs.MAX_LENGTH],
        },
    ),
]
CultivarDescription = Annotated[
    str,
    BeforeValidator(lambda v: v.strip()),
    AfterValidator(SpecManager.get_validation_method(specs, "cultivar_description")),
    # Note: Field used only for annotation, to allow custom error messages.
    Field(
        description=specs.descriptions["cultivar_description"]["field"],
        json_schema_extra={
            "max_length": specs.values["cultivar_description"][Specs.MAX_LENGTH],
        },
    ),
]
CultivarCollectionName = Annotated[
    str,
    BeforeValidator(lambda v: v.strip()),
    AfterValidator(
        SpecManager.get_validation_method(specs, "cultivar_collection_name")
    ),
    # Note: Field used only for annotation, to allow custom error messages.
    Field(
        description=specs.descriptions["cultivar_collection_name"]["field"],
        json_schema_extra={
            "min_length": specs.values["cultivar_collection_name"][Specs.MIN_LENGTH],
            "max_length": specs.values["cultivar_collection_name"][Specs.MAX_LENGTH],
            "pattern": specs.values["cultivar_collection_name"][Specs.PATTERN],
        },
    ),
]
CultivarCollectionDescription = Annotated[
    str,
    BeforeValidator(lambda v: v.strip()),
    AfterValidator(
        SpecManager.get_validation_method(specs, "cultivar_collection_description")
    ),
    # Note: Field used only for annotation, to allow custom error messages.
    Field(
        description=specs.descriptions["cultivar_collection_description"]["field"],
        json_schema_extra={
            "max_length": specs.values["cultivar_collection_description"][
                Specs.MAX_LENGTH
            ],
        },
    ),
]
CultivarCollectionTag = Annotated[
    str,
    BeforeValidator(lambda v: v.strip()),
    AfterValidator(SpecManager.get_validation_method(specs, "cultivar_collection_tag")),
    # Note: Field used only for annotation, to allow custom error messages.
    Field(
        description=specs.descriptions["cultivar_collection_tag"]["field"],
        json_schema_extra={
            "max_length": specs.values["cultivar_collection_tag"][Specs.MAX_LENGTH],
            "pattern": specs.values["cultivar_collection_tag"][Specs.PATTERN],
        },
    ),
]
CultivarCollectionTags = Annotated[
    set[CultivarCollectionTag],
    AfterValidator(
        SpecManager.get_validation_method(specs, "cultivar_collection_tags")
    ),
    # Note: Field used only for annotation, to allow custom error messages.
    Field(
        description=specs.descriptions["cultivar_collection_tags"]["field"],
        json_schema_extra={
            "max_length": specs.values["cultivar_collection_tags"][Specs.MAX_LENGTH],
        },
    ),
]

# ======================================
# Commands
# ======================================


class CultivarCreateCommand(Command):
    """
    Creates a new Cultivar in a CultivarCollection.

    Attributes:
        They follow the same structure as the Cultivar model.
    """

    collection_ref: uuid.UUID
    names: CultivarNames
    key: CultivarKey | None = None
    scientific_name: CultivarScientificName | None = None
    description: CultivarDescription | None = None
    parent_id: uuid.UUID | None = None


class CultivarUpdateCommand(Command):
    """
    Updates an existing Cultivar in a CultivarCollection.
    """

    collection_ref: uuid.UUID
    cultivar_id: uuid.UUID
    names: CultivarNames | None = None
    key: CultivarKey | None = None
    scientific_name: CultivarScientificName | None = None
    description: CultivarDescription | None = None
    parent_id: uuid.UUID | None = None
    remove_parent: bool = False
    """If true, the parent_ref will be deleted regardless of the update value."""
    attributes: CultivarAttributeUpdateCommand | None = None


class CultivarDeleteCommand(Command):
    """
    Deletes a Cultivar from a CultivarCollection.
    """

    collection_ref: uuid.UUID
    cultivar_id: uuid.UUID


class CultivarCollectionCreateCommand(Command):
    """
    Creates a new Cultivar Collection.
    """

    name: CultivarCollectionName
    visibility: CultivarCollectionVisibilityEnum = (
        CultivarCollectionVisibilityEnum.PRIVATE
    )
    description: CultivarCollectionDescription = ""
    tags: CultivarCollectionTags = set()
    parent_ref: uuid.UUID | None = None
    garden_ref: uuid.UUID | None = None


class CultivarCollectionUpdateCommand(Command):
    """
    Updates an existing Cultivar Collection.
    """

    collection_ref: uuid.UUID
    name: CultivarCollectionName | None = None
    visibility: CultivarCollectionVisibilityEnum | None = None
    description: CultivarCollectionDescription | None = None
    tags: CultivarCollectionTags | None = None
    parent_ref: uuid.UUID | None = None
    remove_parent: bool = False
    """If true, the parent_ref will be deleted regardless of the update value."""


class CultivarCollectionDeleteCommand(Command):
    """
    Deletes a Cultivar Collection.
    """

    collection_ref: uuid.UUID


class CultivarCollectionDuplicateCommand(Command):
    """
    Given an existing Cultivar Collection, creates
    a new one with the same data.
    """

    source_collection_ref: uuid.UUID
    garden_ref: uuid.UUID | None = None


class CultivarCollectionMergeCommand(Command):
    """
    Given an existing Cultivar Collection, updates it
    with the data from another collection.
    """

    target_collection_ref: uuid.UUID
    source_collection_ref: uuid.UUID

File: frontend/src/lib/data/cultivar/commands.ts
import { z as zod } from 'zod';
import { useMutation } from '@sveltestack/svelte-query';
import type {
	CultivarCreateCommand,
	CultivarUpdateCommand,
	CultivarDeleteCommand,
	CultivarCollectionCreateCommand,
	CultivarCollectionUpdateCommand,
	CultivarCollectionDeleteCommand,
	CultivarCollectionDuplicateCommand,
	CultivarCollectionMergeCommand
} from '$codegen/types';
import {
	CultivarCollectionCreateCommandVisibility,
	CultivarCollectionUpdateCommandVisibility
} from '$codegen/types';
import {
	cultivarCreateCommandOp,
	cultivarUpdateCommandOp,
	cultivarDeleteCommandOp,
	cultivarCollectionCreateCommandOp,
	cultivarCollectionUpdateCommandOp,
	cultivarCollectionDeleteCommandOp,
	cultivarCollectionDuplicateCommandOp,
	cultivarCollectionMergeCommandOp
} from '$codegen';
import { cultivarFieldSchemas } from './schemas';
import { fi } from '@faker-js/faker';

/** Creates a new cultivar in a collection. */
export const cultivarCreate = {
	schema: zod.object({
		collection_ref: zod.string().uuid(),
		names: cultivarFieldSchemas.cultivar_names,
		key: cultivarFieldSchemas.cultivar_key.optional(),
		scientific_name: cultivarFieldSchemas.cultivar_scientific_name.optional(),
		description: cultivarFieldSchemas.cultivar_description.optional(),
		parent_id: zod.string().uuid().optional()
	}),
	mutation: () => {
		return useMutation(function (data: CultivarCreateCommand) {
			return cultivarCreateCommandOp(data);
		});
	}
};

/** Updates a cultivar. */
export const cultivarUpdate = {
	schema: zod.object({
		collection_ref: zod.string().uuid(),
		cultivar_id: zod.string().uuid(),
		names: cultivarFieldSchemas.cultivar_names.optional(),
		key: cultivarFieldSchemas.cultivar_key.optional(),
		scientific_name: cultivarFieldSchemas.cultivar_scientific_name.optional(),
		description: cultivarFieldSchemas.cultivar_description.optional(),
		parent_id: zod.string().uuid().optional(),
		remove_parent: zod.boolean().optional(),
		attributes: zod.object({
			frost_date_planting_window_profile: zod.object({
				first_frost_window_open:
					cultivarFieldSchemas.first_frost_window_open.optional(),
				first_frost_window_close:
					cultivarFieldSchemas.first_frost_window_close.optional(),
				last_frost_window_open: cultivarFieldSchemas.last_frost_window_open.optional(),
				last_frost_window_close: cultivarFieldSchemas.last_frost_window_close.optional()
			}),
			origin_profile: zod.object({
				transplantable: cultivarFieldSchemas.transplantable.optional()
			}),
			annual_lifecycle_profile: zod.object({
				seed_to_germ: cultivarFieldSchemas.seed_to_germ.optional(),
				germ_to_transplant: cultivarFieldSchemas.germ_to_transplant.optional(),
				germ_to_first_harvest: cultivarFieldSchemas.germ_to_first_harvest.optional(),
				first_to_last_harvest: cultivarFieldSchemas.first_to_last_harvest.optional()
			})
		})
	}),
	mutation: () => {
		return useMutation(function (data: CultivarUpdateCommand) {
			return cultivarUpdateCommandOp(data);
		});
	}
};

/** Deletes a cultivar. */
export const cultivarDelete = {
	schema: zod.object({
		collection_ref: zod.string().uuid(),
		cultivar_id: zod.string().uuid()
	}),
	mutation: () => {
		return useMutation(function (data: CultivarDeleteCommand) {
			return cultivarDeleteCommandOp(data);
		});
	}
};

/** Creates a new cultivar collection. */
export const culviarCollectionCreate = {
	schema: zod.object({
		name: cultivarFieldSchemas.cultivar_collection_name,
		visibility: zod.nativeEnum(CultivarCollectionCreateCommandVisibility).optional(),
		description: cultivarFieldSchemas.cultivar_collection_description.optional(),
		tags: cultivarFieldSchemas.cultivar_collection_tags.optional(),
		parent_ref: zod.string().uuid().optional(),
		garden_ref: zod.string().uuid().optional()
	}),
	mutation: () => {
		return useMutation(function (data: CultivarCollectionCreateCommand) {
			return cultivarCollectionCreateCommandOp(data);
		});
	}
};

/** Updates a cultivar collection. */
export const cultivarCollectionUpdate = {
	schema: zod.object({
		collection_ref: zod.string().uuid(),
		name: cultivarFieldSchemas.cultivar_collection_name.optional(),
		visibility: zod.nativeEnum(CultivarCollectionUpdateCommandVisibility).optional(),
		description: cultivarFieldSchemas.cultivar_collection_description.optional(),
		tags: cultivarFieldSchemas.cultivar_collection_tags.optional(),
		parent_ref: zod.string().uuid().optional(),
		remove_parent: zod.boolean().optional()
	}),
	mutation: () => {
		return useMutation(function (data: CultivarCollectionUpdateCommand) {
			return cultivarCollectionUpdateCommandOp(data);
		});
	}
};

/** Deletes a cultivar collection.. */
export const cultivarCollectionDelete = {
	schema: zod.object({
		collection_ref: zod.string().uuid()
	}),
	mutation: () => {
		return useMutation(function (data: CultivarCollectionDeleteCommand) {
			return cultivarCollectionDeleteCommandOp(data);
		});
	}
};

/** Creates a new cultivar collection by copying another. */
export const cultivarCollectionDuplicate = {
	schema: zod.object({
		source_collection_ref: zod.string().uuid(),
		garden_ref: zod.string().uuid().optional()
	}),
	mutation: () => {
		return useMutation(function (data: CultivarCollectionDuplicateCommand) {
			return cultivarCollectionDuplicateCommandOp(data);
		});
	}
};

/** Merges an existing cultivar collection into another. */
export const cultivarCollectionMerge = {
	schema: zod.object({
		target_collection_ref: zod.string().uuid(),
		source_collection_ref: zod.string().uuid()
	}),
	mutation: () => {
		return useMutation(function (data: CultivarCollectionMergeCommand) {
			return cultivarCollectionMergeCommandOp(data);
		});
	}
};
